### ISSUE TITLE
Tentative plans

### ISSUE THREAD

**Issue by WilCrofter** [WilCrofter/TMS_Long#1]
I'm spending a lot of time on PolyMath 15, so I'll work as time permits but here's a tentative plan:

1. We have API keys and code for pulling data both from FRED and AlphaVantage, so code to pull appropriate data should be quick and easy. I'll probably do that in parallel with:
2. Reading the PDF in hopes that they give enough detail to make the forecast. This means details of their "wavelet filtering methods", and forecast model (linear regression? logistic regression?)
3. If they do give enough detail I'll try to write code to reproduce what they did, specifically Fig 4 of the VoxEu summary. This will be the hard part I think.
4. Economics and other fields (e.g., neuroscience) are having "reproducability crises." Skepticism is called for. Applied types tend to have a fairly shallow grasp of statistics, though enough to game significance tests. They don't necessarily know they are gaming the tests but if they change their models, data preparation, and so on 20 times, they can expect to get a 0.05 (1 in 20) significance reading which is the threshold for publication.

**Comment by WilCrofter** [comment]
Monthly adjusted (i.e. total returns including dividends and splits) for the S&P 500 index (SPX) is available from AlphaVantage as:
 
https://www.alphavantage.co/query?function=TIME_SERIES_MONTHLY_ADJUSTED&symbol=SPX&apikey=******&datatype=csv

My apikey is blocked out of course, but free keys are available [here](https://www.alphavantage.co/support/#api-key). AlphaVantage supports both CSV and JSON data types, but I seem to recall [FRED](https://fred.stlouisfed.org/) which I may have to use for interest rates and spreads supports only CSV, so I've used that here.

**Comment by WilCrofter** [comment]
The full report *does not* give specific wavelet details. I was afraid of that. Instead, they reference a [book](http://staff.washington.edu/dbp/WMTSA/home.html) whose authors were kind enough to provide links to [a variety of software](http://staff.washington.edu/dbp/wmtsa.html), MATLAB, C/C++, etc. so we may be able to reverse engineer what they did.

**Comment by WilCrofter** [comment]
Or maybe [Wavelets.jl](https://github.com/JuliaDSP/Wavelets.jl). It may not be current but can probably be hacked into shape.

![image](https://user-images.githubusercontent.com/2634813/39946669-142d8968-553d-11e8-8e8c-41036b0e558f.png)

**Comment by WilCrofter** [comment]
Wavelets installs and passes its unit tests with warnings about deprecated syntax. I may clone the package and fix the deprecations since Julia 1.0 is down to 5 issues and should soon be released.
```
julia> Pkg.add("Wavelets")
INFO: Cloning cache of Wavelets from https://github.com/JuliaDSP/Wavelets.jl.git
...
julia> Pkg.test("Wavelets")
INFO: Testing Wavelets
WARNING: deprecated syntax "abstract DiscreteWavelet{T}" at /home/wil/.julia/v0.6/Wavelets/src/mod/WT.jl:15.
Use "abstract type DiscreteWavelet{T} end" instead.
...
testing: success
INFO: Wavelets tests passed
```

**Comment by WilCrofter** [comment]
The [Requests](https://github.com/JuliaWeb/Requests.jl) package is deprecated in favor of [HTTP](https://github.com/JuliaWeb/HTTP.jl) which does not look all that well maintained either. Requests is, however, a dependency of [FredData](https://github.com/micahjsmith/FredData.jl) and I have generally used it to pull data from AlphaVantage. I will probably maintain a cloned copy of Requests for the present--I'm familiar with it, having fixed a problem in its makefile a while ago.

**Comment by WilCrofter** [comment]
I've [forked](https://github.com/WilCrofter?tab=repositories) Requests and Wavelets. Hopefully I can maintain the forks well enough to support the present project. I have not forked FredData since it seems reasonably current, the last commit being 3 months ago.

**Comment by alexander-bauer** [comment]
This is a super cursory glance, but it looks like FredData might not make _much_ use of Requests. Is there anything besides [this one code block](https://github.com/micahjsmith/FredData.jl/search?utf8=%E2%9C%93&q=requests&type=)? Maybe it would be easy to patch FredData to use HTTP?

**Comment by WilCrofter** [comment]
Good catch. Requests is actively invoked only twice--the `./scr/get_data.jl` lines below. 

Julia's Request package ultimately (and perhaps unfortunately) depends on [nodejs/http-parser](https://github.com/nodejs/http-parser) whose latest release (almost a year after I raised the issue) at least builds with `gcc 7+`. 
````
~/.julia/v0.6/FredData $ grep -r Requests ./
./src/FredData.jl:using Requests
./src/FredData.jl:import Requests: get
./src/get_data.jl:    obs_json = Requests.json(obs)
./src/get_data.jl:    metadata_json = Requests.json(metadata)
./REQUIRE:Requests
```

**Comment by WilCrofter** [comment]
Julia's HTTP package depends on [MBedTLS](https://github.com/ARMmbed/mbedtls) which looks *very* well maintained (last commit 3 days ago), HTTP itself, less so. HTTP installs, builds, and loads without problems. It fails a few unit tests; the failures don't look serious but the fact that they aren't fixed says something about maintenance. Unfortunately, the failures involve HTTP.client, precisely what we need. So, for instance:
```
julia> using HTTP

julia> r = HTTP.request("GET", "http://httpbin.org/ip"; verbose=3)
ERROR: TypeError: non-boolean (Int64) used in boolean context
```
This is probably a version problem (Int64 ≠ 0 no longer equivalent `true` or something) and would probably be easy to fix. Thinking.

**Comment by WilCrofter** [comment]
Since [JuliaWeb](https://github.com/JuliaWeb) seems poorly maintained I'm inclined to use Julia's intrinsic  [download](https://docs.julialang.org/en/stable/stdlib/file/#Base.download) function whose main drawback is dependency on external tools such as curl, wget or fetch, and the fact that it downloads to a file rather than to memory.

Both AlphaVantage and FRED serve in JSON format, and the [JSON](https://github.com/JuliaIO/JSON.jl) package seems very well maintained. So, I think these two tools are the best choice for data acquisition. Example:
```
julia> download("https://www.alphavantage.co/query?function=TIME_SERIES_MONTHLY_ADJUSTED&symbol=SPX&apikey=myAPIkey&datatype=JSON", "Desktop/file.json")
julia> using JSON
julia> data = JSON.parsefile("Desktop/file.json")
Dict{String,Any} with 2 entries:
  "Monthly Adjusted Time Series" => Dict{String,Any}(Pair{String,Any}("2011-10-…
  "Meta Data"                    => Dict{String,Any}(Pair{String,Any}("1. Infor…
julia> data["Monthly Adjusted Time Series"]
Dict{String,Any} with 220 entries:
  "2011-10-31" => Dict{String,Any}(Pair{String,Any}("6. volume", "98063670000")…
  "2003-10-31" => Dict{String,Any}(Pair{String,Any}("6. volume", "32298500000")…
  "2008-07-31" => Dict{String,Any}(Pair{String,Any}("6. volume", "124980570000"…
 ⋮            => ⋮
julia> data["Monthly Adjusted Time Series"]["2011-10-31"]
Dict{String,Any} with 7 entries:
  "6. volume"          => "98063670000"
  "4. close"           => "1253.3000"
  "5. adjusted close"  => "1253.3000"
  "2. high"            => "1292.6600"
  "1. open"            => "1131.2100"
  "7. dividend amount" => "0.0000"
  "3. low"             => "1074.7700"
```
